query.go

```diff
@@ -43,14 +48,85 @@ type query struct {
 	stopFn stopFn
 }
 
-func (dht *IpfsDHT) runDisjointQueries(ctx context.Context, d int, target string, queryFn queryFn, stopFn stopFn) ([]*query, error) {
-	queryCtx, cancelQuery := context.WithCancel(ctx)
+type lookupResult struct {
+	peers     []peer.ID
+	state     []qpeerset.PeerState
+	completed bool
+}
 
-	numQueriesComplete := 0
-	queryDone := make(chan struct{}, d)
+// runLookup executes the lookup on the target using the given query function and stopping when either the context is
+// cancelled or the stop function returns true (if the stop function is not sticky it is not guaranteed to cause a stop
+// to occur just because it momentarily returns true). Returns the top K peers that have not failed as well as their
+// state at the end of the lookup.
+//
+// After the lookup is complete the query function is run (unless stopped) against all of the top K peers that have not
+// already been successfully queried.
+func (dht *IpfsDHT) runLookup(ctx context.Context, d int, target string, queryFn queryFn, stopFn stopFn) (*lookupResult, error) {
+	// run the query
+	lookupRes, err := dht.runDisjointQueries(ctx, d, target, queryFn, stopFn)
+	if err != nil {
+		return nil, err
+	}
+
+	// query all of the top K peers we've either Heard about or have outstanding queries we're Waiting on.
+	// This ensures that all of the top K results have been queried which adds to resiliency against churn for query
+	// functions that carry state (e.g. FindProviders and GetValue) as well as establish connections that are needed
+	// by stateless query functions (e.g. GetClosestPeers and therefore Provide and PutValue)
+	queryPeers := make([]peer.ID, 0, len(lookupRes.peers))
+	for i, p := range lookupRes.peers {
+		if state := lookupRes.state[i]; state == qpeerset.PeerHeard || state == qpeerset.PeerWaiting {
+			queryPeers = append(queryPeers, p)
+		}
+	}
+
+	if len(queryPeers) == 0 {
+		return lookupRes, nil
+	}
+
+	// return if the lookup has been externally stopped
+	if ctx.Err() != nil || stopFn() {
+		lookupRes.completed = false
+		return lookupRes, nil
+	}
+
+	doneCh := make(chan struct{}, 1)
+	followUpCtx, cancelFollowUp := context.WithCancel(ctx)
+	for _, p := range queryPeers {
+		qp := p
+		go func() {
+			_, _ = queryFn(followUpCtx, qp)
+			doneCh <- struct{}{}
+		}()
+	}
+
+	// wait for all queries to complete before returning, aborting ongoing queries if we've been externally stopped
+processFollowUp:
+	for i := 0; i < len(queryPeers); i++ {
+		select {
+		case <-doneCh:
+			if stopFn() {
+				cancelFollowUp()
+				if i < len(queryPeers)-1 {
```

Comment:
The if statement here just checks that if we're stopping prematurely then the we have not fully completed the lookup + followup. It might be the we call cancel after receiving the very last result so in that case we don't say the lookup + followup are incomplete

renaming the query things to be called lookup will be easier to do after removing the disjoint paths (next PR).

I'm not sure I see the benefit of the extra structs, it seems more confusing than clarifying to me since the real contract exposed to GetClosestPeers, FindProviders, GetValue, etc. is the lookup+followup not the lookup itself. We can couple the lookup + followup results until it becomes inconvenient. If you feel strongly though feel free to put up another PR after it merges so I can see what you want and I'll likely just approve it.